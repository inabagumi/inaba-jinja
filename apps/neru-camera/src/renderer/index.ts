import {
  Color,
  LinearFilter,
  OrthographicCamera,
  PlaneGeometry,
  RGBFormat,
  ShaderMaterial,
  Scene,
  WebGLRenderer,
  Mesh,
  MeshBasicMaterial,
  VideoTexture,
} from 'three';

const fragmentShader = `precision mediump float;
uniform sampler2D texture;
uniform vec3 keyColor;
varying vec2 vUv;
void main(void) {
  vec4 smpColor = texture2D(texture, vUv);
  float diff = length(keyColor - smpColor.rgb);
  if (diff < 0.6) {
    discard;
  } else {
    gl_FragColor = smpColor;
  }
}
`;
const vertexShader = `precision mediump float;
varying vec2 vUv;
void main(void) {
  vUv = uv;
  gl_Position = projectionMatrix * modelViewMatrix * vec4(position, 1.0);
}
`;

export default class Renderer {

  private background: Mesh;
  private camera: OrthographicCamera;
  private overlay: Mesh;
  private renderer: WebGLRenderer;
  private sceen: Scene;

  constructor(canvas: HTMLCanvasElement) {
    this.renderer = new WebGLRenderer({ canvas, preserveDrawingBuffer: true });
    this.sceen = new Scene();

    const { height, width } = this.renderer.getSize();
    this.setCamera(width, height);
  }

  render() {
    this.renderer.render(this.sceen, this.camera);
  }

  setCamera(width: number, height: number) {
    this.camera = new OrthographicCamera(
      width / -2,
      width / 2,
      height / 2,
      height / -2,
      1,
      2000,
    );

    this.camera.position.z = 500;
  }

  setBackground(media: HTMLVideoElement) {
    const { videoHeight, videoWidth } = media;

    this.renderer.setSize(videoWidth, videoHeight, false);
    this.setCamera(videoWidth, videoHeight);

    const texture = new VideoTexture(media);
    texture.minFilter = LinearFilter;
    texture.magFilter = LinearFilter;
    texture.format = RGBFormat;

    const geometry = new PlaneGeometry(videoWidth, videoHeight, 1, 1);
    const material = new MeshBasicMaterial({ map: texture });
    this.background = new Mesh(geometry, material);

    this.sceen.add(this.background);
  }

  setOverlay(media: HTMLVideoElement) {
    const { videoHeight, videoWidth } = media;

    const texture = new VideoTexture(media);
    texture.minFilter = LinearFilter;
    texture.magFilter = LinearFilter;
    texture.format = RGBFormat;

    const geometry = new PlaneGeometry(videoWidth, videoHeight, 1, 1);
    const material = new ShaderMaterial({
      fragmentShader,
      vertexShader,
      transparent: true,
      uniforms: {
        keyColor: {
          type: 'c',
          value: new Color(0x00ff00),
        },
        texture: {
          type: 't',
          value: texture,
        },
      },
    });
    this.overlay = new Mesh(geometry, material);
    this.overlay.scale.x = this.overlay.scale.y = 0.5;

    this.sceen.add(this.overlay);
  }

  start() {
    this.renderer.setAnimationLoop(() => this.render());
  }

  updateTransform(scale: number, x: number, y: number) {
    const { clientHeight, clientWidth, height, width } = this.renderer.domElement;
    const deviation = 100 * ((scale / 0.5) - 1);

    this.overlay.position.x = (x + deviation) / (clientWidth / width);
    this.overlay.position.y = -((y + deviation) / (clientHeight / height));
    this.overlay.scale.x = this.overlay.scale.y = scale;
  }
}
