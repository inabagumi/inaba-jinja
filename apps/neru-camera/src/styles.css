html,
body {
  height: 100%;
  width: 100%;
}

body {
  bottom: 0;
  left: 0;
  margin: 0;
  position: fixed;
  right: 0;
  top: 0;
}
