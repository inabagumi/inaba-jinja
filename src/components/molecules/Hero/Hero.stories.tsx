import React from 'react'
import Hero from './Hero'

export default {
  component: Hero,
  title: 'Hero'
}

export const hero = () => <Hero />
