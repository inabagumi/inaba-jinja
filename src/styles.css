:root {
  --ij-default-font-family: -apple-system, BlinkMacSystemFont,
    Helvetica Neue, Helvetica, Arial, YuGothic, Yu Gothic, sans-serif;
  --ij-serif-font-family: Garamond, Times New Roman, YuMincho, Yu Mincho,
    serif;
}

html {
  box-sizing: border-box;
  font-family: var(--ij-default-font-family);
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-text-size-adjust: none;
}

body {
  background-color: #424242;
  color: #fafafa;
  margin: 0;
  line-height: 2;
}

*,
*::before,
*::after {
  box-sizing: inherit;
}

body,
html,
#__next {
  height: 100%;
}

p {
  margin: 0;
  padding: 0;
}

a {
  color: #f57f17;
}

a:hover {
  color: #ff6f00;
}
