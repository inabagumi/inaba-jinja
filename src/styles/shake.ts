import { keyframes } from '@emotion/react'

const shake = keyframes`
  from {
    transform: translateY(0) rotate(180deg);
  }

  to {
    transform: translateY(20px) rotate(170deg);
  }
`

export default shake
