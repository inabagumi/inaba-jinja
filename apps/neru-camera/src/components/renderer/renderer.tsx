import { makeStyles } from '@material-ui/core/styles'
import { Application, Sprite } from 'pixi.js'
import { Viewport } from 'pixi-viewport'
import React, {
  ReactElement,
  SyntheticEvent,
  forwardRef,
  useCallback,
  useEffect,
  useImperativeHandle,
  useRef,
  useState
} from 'react'
import { Asset } from '../../context/asset-context'
import { ChromaKeyFilter } from '../../filters/ChromaKeyFilter'

const useStyles = makeStyles({
  container: {
    position: 'relative',
    width: '100%'
  },
  preloaded: {
    display: 'none'
  },
  preview: {
    display: 'block',
    height: 'auto',
    width: '100%'
  }
})

export type RefObject = {
  toBlob: (type: string, quality: number) => Promise<Blob>
}

export type Props = {
  asset: Asset
  cameraStream: MediaStream
}

const Renderer = forwardRef<RefObject, Props>(
  ({ asset, cameraStream }, ref): ReactElement => {
    const [viewport, setViewport] = useState<Viewport>()
    const classes = useStyles()

    const previewRef = useRef<HTMLCanvasElement>(null)
    const cameraRef = useRef<HTMLVideoElement>(null)
    const overlayRef = useRef<HTMLVideoElement>(null)

    const setup = useCallback(
      (event: SyntheticEvent<HTMLVideoElement, Event>) => {
        if (!previewRef.current) return

        const background = Sprite.from(event.currentTarget)

        const app = new Application({
          height: background.height,
          preserveDrawingBuffer: true,
          view: previewRef.current,
          width: background.width
        })
        const viewport = new Viewport({
          interaction: app.renderer.plugins.interaction,
          screenHeight: app.screen.height,
          screenWidth: app.screen.width,
          worldHeight: app.screen.height,
          worldWidth: app.screen.width
        })

        viewport
          .drag()
          .pinch()
          .wheel()
          .decelerate()

        app.stage.addChild(background)
        app.stage.addChild(viewport)

        setViewport(viewport)

        if (overlayRef.current) overlayRef.current.load()
      },
      []
    )

    const handleLoadedData = useCallback(
      (event: SyntheticEvent<HTMLVideoElement, Event>) => {
        if (!viewport) return

        const target = event.currentTarget
        const overlay = Sprite.from(target)
        const { keyColor } = target.dataset

        overlay.filters = [new ChromaKeyFilter(keyColor)]
        overlay.anchor.set(0.5)
        overlay.position.set(
          viewport.worldWidth * 0.5,
          viewport.worldHeight * 0.5
        )

        viewport.addChild(overlay)
      },
      [viewport]
    )

    useEffect(() => {
      if (!cameraRef.current) return

      cameraRef.current.srcObject = cameraStream
    }, [cameraStream])

    useImperativeHandle(ref, () => ({
      toBlob: (type = 'image/png', quality = 1.0): Promise<Blob> =>
        new Promise((resolve, reject): void => {
          if (!previewRef.current) {
            reject(new TypeError('Canvas is not mounted.'))
            return
          }

          const canvas = document.createElement('canvas')
          const context = canvas.getContext('2d')

          canvas.width = previewRef.current.width
          canvas.height = previewRef.current.height

          if (!context) {
            reject(new TypeError('Context does not exist.'))
            return
          }

          try {
            context.drawImage(previewRef.current, 0, 0)
            canvas.toBlob(
              blob => {
                if (!blob) {
                  reject(new TypeError('Blob is empty.'))
                  return
                }

                resolve(blob)
              },
              type,
              quality
            )
          } catch (error) {
            reject(error)
          }
        })
    }))

    return (
      <div className={classes.container}>
        <canvas className={classes.preview} ref={previewRef} />

        <video
          className={classes.preloaded}
          crossOrigin="anonymouse"
          muted
          onLoadedData={setup}
          playsInline
          ref={cameraRef}
        />
        <video
          className={classes.preloaded}
          crossOrigin="anonymouse"
          data-key-color={asset.keyColor}
          loop
          muted
          onLoadedData={handleLoadedData}
          playsInline
          preload="none"
          ref={overlayRef}
          src={asset.src}
        />
      </div>
    )
  }
)

export default Renderer
