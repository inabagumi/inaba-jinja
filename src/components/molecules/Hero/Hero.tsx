import { createStyles, makeStyles } from '@material-ui/core/styles'
import clsx from 'clsx'
import Head from 'next/head'
import React, {
  DetailedHTMLProps,
  HTMLAttributes,
  RefForwardingComponent,
  forwardRef
} from 'react'
import mainVisual from './main-visual.jpg'
import webpMainVisual from './main-visual.webp'
import lqipMainVisual from './main-visual@lqip.jpg'

const useStyles = makeStyles(theme => {
  const backgroundColor = 'rgba(0, 0, 0, 0.54)'

  return createStyles({
    content: {
      alignItems: 'center',
      bottom: 0,
      display: 'flex',
      flexDirection: 'column',
      justifyContent: 'center',
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0
    },
    cover: {
      backgroundColor: 'transparent',
      backgroundImage: `url("${lqipMainVisual}")`,
      backgroundPosition: 'center',
      backgroundRepeat: 'no-repeat',
      backgroundSize: 'cover',
      bottom: 0,
      left: 0,
      position: 'absolute',
      right: 0,
      top: 0,

      '&::after': {
        backgroundColor,
        bottom: 0,
        content: '""',
        display: 'block',
        left: 0,
        position: 'absolute',
        right: 0,
        top: 0
      },

      '& img': {
        bottom: 0,
        display: 'block',
        height: '100%',
        left: 0,
        objectFit: 'cover',
        position: 'absolute',
        right: 0,
        top: 0,
        width: '100%'
      }
    },
    root: {
      backgroundColor,
      color: theme.palette.getContrastText(backgroundColor),
      minHeight: '100%',
      position: 'relative'
    }
  })
})

type Props = DetailedHTMLProps<HTMLAttributes<HTMLDivElement>, HTMLDivElement>

const Hero: RefForwardingComponent<HTMLDivElement, Props> = (
  { children, className, ...props },
  ref
) => {
  const classes = useStyles({})

  return (
    <>
      <Head>
        <link
          as="image"
          href={webpMainVisual}
          rel="preload"
          type="image/webp"
        />
      </Head>

      <div className={clsx(classes.root, className)} ref={ref} {...props}>
        <div className={classes.cover}>
          <picture>
            <source srcSet={webpMainVisual} type="image/webp" />

            <img
              alt=""
              height="600"
              role="presentation"
              src={mainVisual}
              width="800"
            />
          </picture>
        </div>

        <div className={classes.content}>{children}</div>
      </div>
    </>
  )
}

export default forwardRef(Hero)
