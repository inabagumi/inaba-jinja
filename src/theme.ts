import deepOrange from '@material-ui/core/colors/deepOrange'
import orange from '@material-ui/core/colors/orange'
import { createMuiTheme } from '@material-ui/core/styles'

export default createMuiTheme({
  palette: {
    primary: deepOrange,
    secondary: orange,
    type: 'light'
  },
  typography: {
    body1: {
      lineHeight: 2
    },
    body2: {
      lineHeight: 2
    },
    fontFamily: [
      'Roboto',
      'Helvetica',
      'Arial',
      'Noto Sans JP',
      'sans-serif'
    ].join(',')
  }
})
