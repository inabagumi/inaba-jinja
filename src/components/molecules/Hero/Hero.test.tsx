/* eslint-env jest */

import { ThemeProvider } from '@material-ui/core/styles'
import React, { FC } from 'react'
import { IntlProvider } from 'react-intl'
import renderer from 'react-test-renderer'
import theme from '../../../theme'
import Hero from './Hero'

const App: FC = ({ children }) => (
  <IntlProvider locale="en">
    <ThemeProvider theme={theme}>{children}</ThemeProvider>
  </IntlProvider>
)

describe('Hero', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <App>
          <Hero />
        </App>
      )
      .toJSON()

    expect(tree).toMatchSnapshot()
  })
})
