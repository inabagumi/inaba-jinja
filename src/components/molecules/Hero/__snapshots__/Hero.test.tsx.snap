// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Hero renders correctly 1`] = `
<div
  className="makeStyles-hero-1"
>
  <SvgMock
    aria-label="因幡神社"
    className="makeStyles-root-3 makeStyles-logo-2"
    focusable="false"
    role="img"
  />
</div>
`;
